### PR TITLE
appModules/SearchUI: add support for announcing suggestion count in Windows 10 Version 1909 and later

### LINKED ISSUE #10341: Windows 10/Windows Search: add support for File Explorer embedded search experience in Version 1909 and later
Hi,

Another one in a series of issues on Windows 10/Windows Search:

### Background:
Windows 10 Version 1909 (likely to be called October 2019 Update) and later includes online search capability in File Explorer. This feature, powered by redesigned Windows Search, is used to allow users to search OneDrive content. As the search progresses, a live region informs result count. Because of the way search box is handled, braille will initially display "suggestions" instead of displaying this live region.

### Is your feature request related to a problem? Please describe.
As noted above, a live region text is used to announce results count. Unfortunately this region isn't labeled, and the text itself is the first child of this region, necessitating NVDA descending one level to announce this live region.

Also, when using braille, the word "suggestions" is brailed before the live region is displayed. This is caused by search box recognition code in SearchUI app module where Start menu search field is recognized as a pure UIA search field. Thus suggestionsOpened event must be modified in sTart menu search field to not braille suggestions message (a separate issue but is grouped under this one to provide context).

### Describe the solution you'd like
Add support for Version 1909 File Explorer/Windows Search facility by:

1. For the live region: give it the name of the child object so live region change event can function properly.
2. Add a custom version of suggestionsOpened event for Start menu search field so braille won't have to say "suggestions".
3. Optionally, rename searchui.StartMenuSearchField to something else (later).

### Describe alternatives you've considered
Leave it as is.

### Additional context
The fix described above is part of Windows 10 App Essentials add-on and is being tested. Also, it would be ideal to see this feature implemented by 2019.3 or shortly after. Also, in 20H1, Windows Search executable has been renamed to SearchApp.exe, so we need to take care of that by 2020.1 release.

Thanks.

### PR DESCRIPTION
**Pull request by josephsl**
### Link to issue number:
Fixes #10341 

### Summary of the issue:
In Windows 10 Version 1909, File Explorer gains ability to search OneDrive content via Windows Search, with a live region used to provide suggestion count announcement.

### Description of how this pull request fixes the issue:
In Windows 10 Version 1909 (build 18363) and later, File Explorer gains ability to search OneDrive content. This is powered by an embedded Windows Search feature (SearchUI.exe) with a live region announcing suggestion count. This live region has no label, but the child object does, thus coerce the name of this object to be the name of its child, which then allows live region change event to work.

### Testing performed:
Tested in Version 1909 (build 18363) and via Windows 10 App Essentials add-on.

### Known issues with pull request:
None

### Change log entry:
Possibly either new features or changes:

In Windows 10 Version 1909 (change this wording to marketing name once 1909 comes out), NVDA will announce suggestion count when performing searches in File Explorer. (#10231

### CODE REVIEW

**Comment by leonardder** [comment]
Is this somehow related to #8472?

**Comment by josephsl** [comment]
Hi, I can see a connection – let me test that PR to see if this one can be rephrased as such. Thanks.

**Comment by josephsl** [comment]
Hi, this is different than #8466, as there is no “alert” associated with this one. Thanks.

**Comment by josephsl** [comment]
Hi,

It is ready for another round of review after 2019.3 is released. Thanks.

**Comment by josephsl** [comment]
Hi,

Indeed, 2019.3 was released.

Thanks.

**Comment by josephsl** [comment]
Hi,

Update: I haven't forgotten about this PR - it's just that more important pull requests came along and developers are going through backlog.

Thanks.

**Comment by codeofdusk** [comment]
Any updates on this PR?

**Comment by josephsl** [comment]
Hi, I might resurrect this for 2021.2 – see a discussion between I and @JCSTeh for some thoughts about this. In fact, I just committed changes to Windows 10 App Essentials that might help with this PR. Note that announcing suggestion count applies to apps other than SearchUI/SearchApp (Windows Search) because these apps do report suggestion count (navigator object moves to suggestions list which does report position info). Thanks.

**Comment by josephsl** [comment]
Hi,
Never mind about the above comment, as I thought it was #7330. The PR is indeed ready to go - I'd wait for 2021.2 cycle begins before requesting another formal review.

Thanks.

File: source/appModules/searchui.py [hunk 1]
```diff
@@ -18,6 +18,15 @@ class StartMenuSearchField(SearchField):
 
 class AppModule(appModuleHandler.AppModule):
 
+	def event_NVDAObject_init(self, obj):
+		if isinstance(obj, UIA):
+			# #10341: Build 18363 introduces modern search experience in File Explorer.
```
Comment (feerrenrut) [inline source/appModules/searchui.py:23]: If this is only available in Build 18363 and later, should this be protected by a check for the windows version? If not, please describe in the comment how this code handles earlier versions. I assume that the `UIAAutomationId` wont be `suggestionCountForNarrator` in earlier builds of Windows?

File: source/appModules/searchui.py [hunk 2]
```diff
@@ -27,7 +36,7 @@ def chooseNVDAObjectOverlayClasses(self,obj,clsList):
 			except ValueError:
 				pass
 		elif isinstance(obj, UIA):
-			if obj.UIAElement.cachedAutomationID == "SearchTextBox":
+			if obj.UIAAutomationId == "SearchTextBox":
```
Comment (feerrenrut) [inline source/appModules/searchui.py:39]: Perhaps this is only ever true in post 18363 builds?

**Comment by josephsl** [comment]
Hi, modern search integration was introduced in 18363, and prior to that, the search UI is different (I’ll add a check for windows releases just to make sure). Thanks.

**Comment by feerrenrut** [comment]
Ok. But the problem I want to solve is:
"From the perspective of a code reader, how do they know that this code will only run on the intended versions of windows?"

**Comment by josephsl** [comment]
Hi, one simple solution I think (which I will do before I sleep) is specifying on which Windows 10 version the code will run on (winVersion.WIN10_1909, to be exact). Thanks.

File: source/appModules/searchui.py (hunk 2 above)
Comment (feerrenrut) [reply source/appModules/searchui.py:44]: Can you verify whether this change is applicable to earlier builds? It may need the same windows version check.

**Comment by josephsl** [comment]
Hi, no need to change this, as SearchUI was introduced in Windows 10, although I think a header comment should clarify this fact (@MichaelDCurran had an adventure with this app module in particular back in 2015 because Windows Search kept freezing for some odd reason). Thanks,

**Comment by josephsl** [comment]
Hi,

A component of #12585 because searchapp.exe was renamed to searchhost.exe in Windows 11, complete with exact same live region implementation as this PR. Therefore, once this PR is erviewed and merged, it can be applied easily to Windows 11 by introducing an alias app module for searchhost.exe that will import from searchui.py.

Thanks.

**Comment by lukaszgo1** [comment]
@feerrenrut  It looks like all changes that you've requested has been addressed. Would it be possible to re-review this?

**Comment by feerrenrut** [comment]
How to test: 
- Open explorer window `Windows + e`
- Press F3 to focus search
- Type something. Note "10 suggestions available, press down arrow key to access results" announced, this is new.
- Press down arrow, file name and details should be announced

**Review by feerrenrut — APPROVED** [review]
Thanks @josephsl

**Pull request merged**